core/domain/classifier_services.py

```diff
@@ -0,0 +1,338 @@
+
+        statez = {
+            'updates': 0,
+            'computes': 0
+        }
+
+        for d in doc_ids:
+            doc, labels = self._get_record(d)
+            for p, w in enumerate(doc):
+                l = self._l_dp[d][p]
+
+                self._update_counting_matrices(d, w, l, False)
+
+                # Gibbs update of labels
+                coeff_a = 1.0 / (
+                    self._c_dl[d].sum() + self._num_labels * self._alpha)
+                coeff_b = 1.0 / (
+                    self._c_lw.sum(axis=1) + self._num_words * self._beta)
+                prob_l = (
+                    labels *
+                    coeff_a * (self._c_dl[d] + self._alpha) *
+                    coeff_b * (self._c_lw[:, w] + self._beta))
+                new_l = numpy.random.multinomial(
+                    1, prob_l / prob_l.sum()).argmax()
+
+                statez['computes'] += 1
+                if l != new_l:
+                    statez['updates'] += 1
+
+                self._l_dp[d][p] = new_l
+                self._update_counting_matrices(d, w, new_l)
+
+        return statez
+
+    def _get_probability(self, d):
+        """Returns the probability of a doc having a label."""
+        probs_i = self._c_dl[d] + (self._b_dl[d] * self._alpha)
+        probs_i = probs_i / probs_i.sum(axis=0)[numpy.newaxis]
+        return probs_i
+
+    def _get_probability_with_label(self, d):
+        """Returns a dict that maps human-readable labels to the probability
+        that the doc has the label."""
+        probs_i = self._get_probability(d)
+        probs_l = {}
+        for label, l in self._label_to_id.iteritems():
+            probs_l[label] = probs_i[l]
+        return probs_l
+
+    def _predict(self, d, threshold=0.5):
+        """Calculates prediction data for a doc.
+        Returns a tuple of (predicted label, prediction score)."""
+        probs_l = self._get_probability_with_label(d)
+        default_prob = probs_l[self._DEFAULT_LABEL]
+        default_prob_comp = 1.0 - default_prob
+        for label in sorted(probs_l, key=probs_l.get, reverse=True):
+            prob_l = probs_l[label]
+            if (label != self._DEFAULT_LABEL and
+                    prob_l / default_prob_comp > threshold):
+                return label, prob_l / default_prob_comp
+        return self._DEFAULT_LABEL, default_prob / default_prob_comp
+
+    def _validate_label(self, label):
+        """Checks that labels don't begin with underscore."""
+        if label[0] == '_':
+            raise Exception('Label %s cannot begin with underscore.' % label)
+
+    def _parse_examples(self, examples):
+        """Splits examples into docs (split on spaces) and labels.
+        Example format:
+        [
+            ['doc1', ['label1']],
+            ['doc2', ['label2']],
+            etc
+        ]
+        """
+        docs = []
+        labels = []
+        for example in examples:
+            doc = example[0].split()
```

Comment:
Good point - I wanted normalization to happen separate of the classifier because I think that code can be reused.